File: src/RayTracing/TypeScript/Objects.ts
import Ray from "./Ray";
import { SolveQuadratic } from "./Solver";
import Vec from "./Vector3";

type Material = {
  colour: [number, number, number],
  specular: number,
};

class Sphere {
  position: [number, number, number];
  radius: number;
  material: Material;

  constructor(position: [number, number, number], radius: number, material: Material) {
    this.position = position;
    this.radius = radius;
    this.material = material;
  }

  public intersect(ray: Ray): number | null {
    // working out in whiteboard
    const newOrigin = Vec.sub(ray.origin, this.position);

    const a: number = 1;
    const b: number = 2 * Vec.dot(ray.direction, newOrigin);
    const c: number = Vec.dot(newOrigin, newOrigin) - (this.radius ** 2);

    const solutions = SolveQuadratic(a, b, c);

    // doesn't intersect
    if (solutions === null) return null;

    // just touches at 1 point
    if (!Array.isArray(solutions)) return solutions;

    // return the minimum
    return solutions[0] < solutions[1] ? solutions[0] : solutions[1];
  }

  public normalAtPoint(point: [number, number, number]): [number, number, number] {
    // simple circle stuff
    const normal = Vec.normalize(Vec.sub(point, this.position));
    return normal;
  }
}

// eslint-disable-next-line import/prefer-default-export
export { Sphere };

File: src/RayTracing/rs-ray-tracing/src/objects.rs
use serde::{Serialize, Deserialize};

use crate::vec3::Vec3;
use crate::ray::Ray;
pub use crate::solver::solve_quadratic;

#[derive(Clone, Copy, Serialize, Deserialize)]
pub struct Material {
  pub colour: (f64, f64, f64),
  pub specular: f64,
}

#[derive(Clone, Copy, Serialize, Deserialize)]
pub struct Sphere {
  pub center: Vec3,
  pub radius: f64,
  pub material: Material,
}

impl Sphere {
  pub fn intersect (&self, ray: &Ray) -> Option<f64> {
    // working out in whiteboard
    let new_origin = ray.origin - self.center;

    let a = 1.;
    let b = 2. * ray.direction.dot(new_origin);
    let c = new_origin.dot(new_origin) - self.radius.powi(2);

    let solution = solve_quadratic(a, b, c);

    match solution {
        Some(solution) => {
          if solution.0 < solution.1 {
            Some(solution.0)
          } else {
            Some(solution.1)
          }
        }
        None => None
    }
  }

  pub fn normal_at_point (&self, point: Vec3) -> Vec3 {
    // simple circle stuff
    (point - self.center).normalize()
  }
}

pub enum Light {
  Direction { intensity: (f64, f64, f64), direction: Vec3},
  Point { intensity: (f64, f64, f64), position: Vec3},
}

impl Light {
  pub fn intensity(&self, _point: Vec3) -> (f64, f64, f64) {
    match self {
      Light::Direction { intensity, direction: _ } => *intensity,
      Light::Point { intensity, position: _ } => *intensity,
    }
  }

  pub fn direction(&self, point: Vec3) -> Vec3 {
    match self {
      Light::Direction { intensity: _, direction } => *direction,
      Light::Point { intensity: _, position } => *position - point,
    }
  }
}
